### PR TITLE
trace payment_continue and payment mods

### PR DESCRIPTION
**Pull request by JssDWt**

### CODE REVIEW

**Comment by cdecker** [comment]
Thanks @JssDWt, I've been wanting to instrument this part of the code for quite a while, and never had the muse to get started. I added #7575 on top of your change to instrument all JSON-RPC calls that plugins start, and we can likely also instrument all entrypoints to the plugin as well (rpcmethod, hook and notifications).

File: plugins/libplugin-pay.c [hunk 1]
```diff
@@ -2266,19 +2268,27 @@ void payment_continue(struct payment *p)
...
 	void *moddata;
+
+	trace_span_start("payment_continue", p);
 	/* If we are in the middle of calling the modifiers, continue calling
 	 * them, otherwise we can continue with the payment state-machine. */
 	p->current_modifier++;
 	mod = p->modifiers[p->current_modifier];
 
 	if (mod != NULL) {
+		char str[2];
```
Comment (cdecker) [inline plugins/libplugin-pay.c:2279]: We generally don't allow the use of `sprintf` as it isn't safe (it takes a buffer it doesn't know the size of). Please switch this to `snprintf` which errors out if the buffer is too small for the format. Or better `tal_fmt(tmpctx, "%d", num);` will do all of it at once, including fixing the buffer overflow should we ever have 100+ modifiers 😋

**Review by cdecker — CHANGES_REQUESTED** [review]

File: plugins/libplugin-pay.c (hunk 1 above)
Comment (JssDWt) [reply plugins/libplugin-pay.c:2279]: I saw you already corrected that in #7575.
Should I fix that in this commit? Or close this PR and go with your PR there?

Comment (cdecker) [reply plugins/libplugin-pay.c:2279]: However we want to merge them. We can merge #7575 and your commit should still be credited to you 🤗 Or we fix up this one, merge this and then my follow-up PR.

By the way, if you check the allow maintainer checkbox when opening a PR you allow us to amend, fix up and fine tune PRs in situ 😉

**Comment by JssDWt** [comment]
@cdecker I moved the commits around a little bit. Only tracing payment_continue now, because under some circumstances it seemed that the traces got intertwined. Also incorporated your suggestion [here](https://github.com/ElementsProject/lightning/pull/7556#discussion_r1719490112)

File: plugins/libplugin-pay.c (hunk 1 above)
Comment (JssDWt) [reply plugins/libplugin-pay.c]: Fixed it up in this one. Although this one includes your commit now :-S

**Comment by JssDWt** [comment]
Rebased. And also fixed a bug where the `error` field would be `false` when there was an error and vice versa.

**Comment by rustyrussell** [comment]
Pretty!  I missed this on the milestone, fixed now.

Rebasing now.

**Comment by rustyrussell** [comment]
This is confusing.  https://github.com/ElementsProject/lightning/pull/7575 says it builds on this, this says it's now based on that.  And I can't force push to this branch to rebase.

So I'm going to add to the confusion by opening another PR!

**Pull request closed without merging**